### PR TITLE
[Merged by Bors] - feat: A result on composition of lift of MonoidLocalization

### PR DESCRIPTION
**Pull request by XavierXarles**
Added a result on composition of lift of the localization map on Monoids. Modified some cases'. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by riccardobrasca** [comment]
Is the tag "Bug" wanted?

**Comment by XavierXarles** [comment]
> Is the tag "Bug" wanted?

not anymore

File: Mathlib/GroupTheory/MonoidLocalization.lean [hunk 1]
```diff
@@ -1087,27 +1085,47 @@ theorem lift_id (x) : f.lift f.map_units x = x :=
...
-  apply k.eq_of_eq f.map_units
-  rw [k.toMap.map_mul, k.toMap.map_mul, ← sec_spec, mul_assoc, lift_spec_mul]
-  repeat' rw [← k.toMap.map_mul]
-  apply f.eq_of_eq k.map_units
-  repeat' rw [f.toMap.map_mul]
-  rw [sec_spec', ← hx]
-  ac_rfl
+have aux : k.lift f.map_units (f.lift k.map_units z) = f.lift f.map_units z :=
+  congrFun (congrArg DFunLike.coe (lift_comp_lift_eq f k f.map_units)) z
+rw [aux]
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/MonoidLocalization.lean:1127]: This should be indented two spaces

File: Mathlib/GroupTheory/MonoidLocalization.lean [hunk 2]
```diff
@@ -1087,27 +1085,47 @@ theorem lift_id (x) : f.lift f.map_units x = x :=
...
+`k : M →+ Q` for a Submonoid `T ⊆ M`, such that `S ≤ T`, and we have
+`l : M →+ A`, the composition of the induced map `f.lift` for `k` with
+the induced map `k.lift` for `l` is equal to the  induced map `f.lift` for `l`"]
+theorem lift_comp_lift {T : Submonoid M} (hST: S ≤ T) {Q : Type*} [CommMonoid Q]
+    (k : Submonoid.LocalizationMap T Q) {A : Type*} [CommMonoid A]{l : M →* A}
+    (hl : ∀ w : T, IsUnit (l w)) :
+    (k.lift hl).comp (f.lift (fun x => map_units k ⟨ _ , hST (SetLike.coe_mem x)⟩)) =
+    f.lift (fun x ↦ hl ⟨ _ , hST (SetLike.coe_mem x)⟩) := by
+  have hlS : ∀ (y : S), IsUnit (l ↑y) := (fun x ↦ hl ⟨ _ , hST (SetLike.coe_mem x)⟩)
+  have hkS : ∀ (y : S), IsUnit (k.toMap ↑y) := (fun x => map_units k ⟨ _ , hST (SetLike.coe_mem x)⟩)
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/MonoidLocalization.lean:1103]: nit:
```suggestion
  have hlS (y : S) : IsUnit (l ↑y) := hl ⟨ _ , hST (SetLike.coe_mem y)⟩
  have hkS (y : S) : IsUnit (k.toMap ↑y) := map_units k ⟨ _ , hST (SetLike.coe_mem y)⟩
```

File: Mathlib/GroupTheory/MonoidLocalization.lean [hunk 3]
```diff
@@ -1087,27 +1085,47 @@ theorem lift_id (x) : f.lift f.map_units x = x :=
...
+  let j := (k.lift hl).comp (f.lift hkS)
+  suffices aux1: j.comp f.toMap = l
+  exact (lift_unique f hlS (fun x ↦ congrFun (congrArg DFunLike.coe aux1) x)).symm
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/MonoidLocalization.lean:1106]:
```suggestion
  suffices aux1 : j.comp f.toMap = l by
    exact (lift_unique f hlS (fun x ↦ congrFun (congrArg DFunLike.coe aux1) x)).symm
```

File: Mathlib/GroupTheory/MonoidLocalization.lean [hunk 4]
```diff
@@ -1087,27 +1085,47 @@ theorem lift_id (x) : f.lift f.map_units x = x :=
...
+  have aux2 : j.comp f.toMap = (k.lift hl).comp k.toMap:= by
+    rw[← lift_comp f hkS, @MonoidHom.comp_assoc]
+  rw [aux2]
+  exact lift_comp k hl
+
+@[to_additive]
+theorem lift_comp_lift_eq {Q : Type*} [CommMonoid Q] (k : Submonoid.LocalizationMap S Q)
+    {A : Type*} [CommMonoid A]{l : M →* A} (hl : ∀ w : S, IsUnit (l w)) :
+    (k.lift hl).comp (f.lift (fun _ => map_units k _) ) = f.lift hl :=
+  lift_comp_lift f (le_refl S) k hl
```
Comment (alreadydone) [inline Mathlib/GroupTheory/MonoidLocalization.lean:1116]:
```suggestion
/-- Given Localization maps `f : M →* N` for a Submonoid `S ⊆ M` and
`k : M →* Q` for a Submonoid `T ⊆ M`, such that `S ≤ T`, and we have
`l : M →* A`, the composition of the induced map `f.lift` for `k` with
the induced map `k.lift` for `l` is equal to the  induced map `f.lift` for `l`. -/
@[to_additive
    "Given Localization maps `f : M →+ N` for a Submonoid `S ⊆ M` and
`k : M →+ Q` for a Submonoid `T ⊆ M`, such that `S ≤ T`, and we have
`l : M →+ A`, the composition of the induced map `f.lift` for `k` with
the induced map `k.lift` for `l` is equal to the  induced map `f.lift` for `l`"]
theorem lift_comp_lift {T : Submonoid M} (hST: S ≤ T) {Q : Type*} [CommMonoid Q]
    (k : LocalizationMap T Q) {A : Type*} [CommMonoid A] {l : M →* A}
    (hl : ∀ w : T, IsUnit (l w)) :
    (k.lift hl).comp (f.lift (map_units k ⟨_, hST ·.2⟩)) =
    f.lift (hl ⟨_, hST ·.2⟩) := .symm <|
  lift_unique _ _ fun x ↦ by rw [← MonoidHom.comp_apply, MonoidHom.comp_assoc, lift_comp, lift_comp]

@[to_additive]
theorem lift_comp_lift_eq {Q : Type*} [CommMonoid Q] (k : LocalizationMap S Q)
    {A : Type*} [CommMonoid A] {l : M →* A} (hl : ∀ w : S, IsUnit (l w)) :
    (k.lift hl).comp (f.lift k.map_units) = f.lift hl :=
  lift_comp_lift f le_rfl k hl
```

File: Mathlib/GroupTheory/MonoidLocalization.lean [hunk 5]
```diff
@@ -1087,27 +1085,47 @@ theorem lift_id (x) : f.lift f.map_units x = x :=
...
-  rw [k.toMap.map_mul, k.toMap.map_mul, ← sec_spec, mul_assoc, lift_spec_mul]
-  repeat' rw [← k.toMap.map_mul]
-  apply f.eq_of_eq k.map_units
-  repeat' rw [f.toMap.map_mul]
-  rw [sec_spec', ← hx]
-  ac_rfl
+have aux : k.lift f.map_units (f.lift k.map_units z) = f.lift f.map_units z :=
+  congrFun (congrArg DFunLike.coe (lift_comp_lift_eq f k f.map_units)) z
+rw [aux]
+exact lift_id f z
```
Comment (alreadydone) [inline Mathlib/GroupTheory/MonoidLocalization.lean:1128]:
```
    k.lift f.map_units (f.lift k.map_units z) = z :=
  (DFunLike.congr_fun (lift_comp_lift_eq f k f.map_units) z).trans (lift_id f z)
```

**Review by alreadydone — APPROVED** [review]

File: Mathlib/GroupTheory/MonoidLocalization.lean [hunk 6]
```diff
@@ -444,14 +444,12 @@ protected irreducible_def smul [SMul R M] [IsScalarTower R M M] (c : R) (z : Loc
...
-        cases' b' with b' hb'
+        let ⟨b, hb⟩ := b
+        let ⟨b', hb'⟩ := b'
         rw [r_eq_r'] at h ⊢
-        cases' h with t ht
+        let ⟨t, ht⟩ := h
         use t
         dsimp only [Subtype.coe_mk] at ht ⊢
--- TODO: this definition should take `SMulCommClass R M M` instead of `IsScalarTower R M M` if
--- we ever want to generalize to the non-commutative case.
```
Comment (alreadydone) [inline Mathlib/GroupTheory/MonoidLocalization.lean:454]: Why remove these? I think they refer to `[IsScalarTower R M M]` in the statement.

File: Mathlib/GroupTheory/MonoidLocalization.lean [hunk 7]
```diff
@@ -1087,27 +1087,37 @@ theorem lift_id (x) : f.lift f.map_units x = x :=
...
+/-- Given Localization maps `f : M →* N` for a Submonoid `S ⊆ M` and
+`k : M →* Q` for a Submonoid `T ⊆ M`, such that `S ≤ T`, and we have
+`l : M →* A`, the composition of the induced map `f.lift` for `k` with
+the induced map `k.lift` for `l` is equal to the  induced map `f.lift` for `l`. -/
+@[to_additive
+    "Given Localization maps `f : M →+ N` for a Submonoid `S ⊆ M` and
+`k : M →+ Q` for a Submonoid `T ⊆ M`, such that `S ≤ T`, and we have
+`l : M →+ A`, the composition of the induced map `f.lift` for `k` with
+the induced map `k.lift` for `l` is equal to the  induced map `f.lift` for `l`"]
+theorem lift_comp_lift {T : Submonoid M} (hST: S ≤ T) {Q : Type*} [CommMonoid Q]
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/MonoidLocalization.lean:1099]:
```suggestion
theorem lift_comp_lift {T : Submonoid M} (hST : S ≤ T) {Q : Type*} [CommMonoid Q]
```

File: Mathlib/GroupTheory/MonoidLocalization.lean [hunk 8]
```diff
@@ -1087,27 +1087,37 @@ theorem lift_id (x) : f.lift f.map_units x = x :=
...
+    (hl : ∀ w : T, IsUnit (l w)) :
+    (k.lift hl).comp (f.lift (map_units k ⟨_, hST ·.2⟩)) =
+    f.lift (hl ⟨_, hST ·.2⟩) := .symm <|
+  lift_unique _ _ fun x ↦ by rw [← MonoidHom.comp_apply,
+    MonoidHom.comp_assoc, lift_comp, lift_comp]
+
+@[to_additive]
+theorem lift_comp_lift_eq {Q : Type*} [CommMonoid Q] (k : Submonoid.LocalizationMap S Q)
+    {A : Type*} [CommMonoid A]{l : M →* A} (hl : ∀ w : S, IsUnit (l w)) :
+    (k.lift hl).comp (f.lift (fun _ => map_units k _) ) = f.lift hl :=
```
Comment (alreadydone) [inline Mathlib/GroupTheory/MonoidLocalization.lean:1110]:
```suggestion
theorem lift_comp_lift_eq {Q : Type*} [CommMonoid Q] (k : LocalizationMap S Q)
    {A : Type*} [CommMonoid A] {l : M →* A} (hl : ∀ w : S, IsUnit (l w)) :
    (k.lift hl).comp (f.lift k.map_units) = f.lift hl :=
```

**Comment by urkud** [comment]
Please mention which `lift` you're talking about either in the title or in the commit message.

**Comment by alreadydone** [comment]
> Please mention which `lift` you're talking about either in the title or in the commit message.

Can you address this as well?

**Comment by XavierXarles** [comment]
> > Please mention which `lift` you're talking about either in the title or in the commit message.
> 
> Can you address this as well? You should be able to edit your topmost post, which will become the commit message.

Is that sufficient?

**Comment by alreadydone** [comment]
Oh actually you already changed the title before my comment, which would be sufficient if I noticed. Thanks!

maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**